### DISCUSSION TITLE
Pause/Resume simulation while waiting for files (for coupling with another software) [PropertyReadFile]

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#26552] [Q&A General] [answered]
Hi guys,

just wondering if there anyway that I can pause MOOSE simulation while waiting for files that will be generated (but not yet) on a folder? This is probably related to `PropertyReadFile` object as well, since it now can read multiple files from a single object/block.

The idea will be, we want to do a close coupling between MOOSE (hydro) and our Abaqus (mech) via damage files. The files that we want to exchange with MOOSE can be read via `PropertyReadFile`, we have done loose coupling before, where we generated 100+ of damage files and let MOOSE-hydro read it at each timestep.

I will break down the step, and please let me know if it's possible

- run Abaqus mech > produce damage file for t=1 > pause (done)
- run MOOSE hydro > read damage file for t= 1 > run simulation > produce pwp_1.csv > pause (maybe at timestep_end)
- resume Abaqus mech > taking pwp from MOOSE > produce damage file for t=2 > pause (done)
- resume MOOSE hydro > read damage file for t= 2 > run simulation > produce pwp_2.csv >  > pause
- repeat

The pause/resume part in Abaqus is already done by us, but not yet with MOOSE

Problems I see so far are: 

- **even we know the total number of the files and all the file names, `PropertyReadFile` wont let us pass the initial setup since all the files are not there**.
- **We dont know how to pause/resume MOOSE simulation (not even sure if this is possible)**

If you have any idea on how to overcome this issue please let me know.

Kind regards,
Traiwit

@GiudGiud I tag you just in case you know if this possible OR you know someone that has done this before 👍

**Comment by GiudGiud** [comment]
@friedmud might want to respond since that could be a use for the MooseServer?

**Reply by Traiwit** [reply]
Hi @GiudGiud

I checked your new `PropertyReadFile` it almost works in such a way we want

first of all it does not require all the files to be in the folder upfront.

is there a way to modify it such that instead of splitting out the error because of a particular file isnt there at a particular timestep


For example the error below occured at the begining of t=3 (looking for damage 3) when the damage_03.csv isnt there, but the first 2 timesteps were fine.

```
Time Step 3, time = 0.00015, dt = 5e-05

*** ERROR ***
Unable to open file "/home/moose/Hydro_Krw_test/damage_03.csv". Check to make sure that it exists and that you have read permission.

```

we want to make an infinate while loop and keep checking (idk every 5 sec) if the file is there, once it's there break the loop then move on into the solving stage of that timestep.

do you think MOOSE can do this?

Kind regards,
Traiwit

**Reply by Traiwit** [reply, marked as answer]
i think i got it

```
void PropertyReadFileCoupled::initialize()
{
  // Since we read at construction, no need to re-read the file on first initialize
  if (!_initialize_called_once)
  {
    _initialize_called_once = true;
    return;
  }

  // Set then read new file, only if we have not reached the last file
  if (_current_file_index < _prop_file_names.size())
  {
    bool file_available = false;
    while(!file_available)
    {
      std::ifstream test(_prop_file_names[_current_file_index]);
      if(test.good())
      {
        // File exists and is readable
        file_available = true;
        test.close();
      }
      else
      {
        // File does not exist or is not readable, wait for 5 seconds
        std::this_thread::sleep_for(std::chrono::seconds(5));
      }
    }

    _reader.setFileName(_prop_file_names[_current_file_index]);
    readData();
  }
  else if (_prop_file_names.size() > 1 && _current_file_index == _prop_file_names.size())
  {
    mooseInfo("Last file specified has been read. The file will no longer be updated.");
  }
}
```
thank you @josebastiase again for the idea!

**Reply by GiudGiud** [reply]
Yup this looks like a good solution to me.

One comment to make this friendlier to the user here would be to add a perf graph live section.
That way if MOOSE starts waiting a long time for that file, the perf graph will start printing a message, then dot dot dots every few seconds, to let you know where and why it is waiting

**Comment by josebastiase** [comment]
Yeah I done that... you dont need to stop/pause the simulations, you have to make MOOSE "wait" while ABAQUS is simulating and vice versa. That way, you wont loose time and resources restarting after every time step. You can use an UserObject for that and the ABAQUS python interface.

For the files initial time step, I just created "fake" data files with the names of the coupling files so `PropertyReadFile` can pass.

I hope this makes sense ;)

**Reply by Traiwit** [reply]
Hi @josebastiase 

Thank you for your reply! good to know that it's achievable.

Could you please elaborate:

> You can use an UserObject for that and the ABAQUS python interface.

which `UserObject` did you use to make MOOSE 'wait'? could you please guide me on this or share if you have made one.

Kind regards,
Traiwit

**Comment by josebastiase** [comment]
Unfortunately, I dont have access to the files anymore nor to an ABAQUS licence. But you have to write the UserObject yourself with a bit of C+ and do something like: while ABAQUS is simulating - wait.

if you cannot picture this drop me an email ;)

**Reply by Traiwit** [reply]
Hi Jose,

Just sent you an email, I looked around MOOSE site, couldn't find anything related to 'wait' at all

Cheers,
Traiwit
